test/TestImportIncludeRole.py

```diff
@@ -55,56 +57,68 @@
 '''
 
 
-class TestImportIncludeRole(unittest.TestCase):
-    def setUp(self):
-        rulesdir = os.path.join('lib', 'ansiblelint', 'rules')
-        self.rules = RulesCollection([rulesdir])
-
-        # make dir and write role tasks to import or include
-        self.play_root = tempfile.mkdtemp()
-        role_path = os.path.join(self.play_root, 'test-role', 'tasks')
-        os.makedirs(role_path)
-        with open(os.path.join(role_path, 'main.yml'), 'w') as f_main:
-            f_main.write(ROLE_TASKS_MAIN)
-        with open(os.path.join(role_path, 'world.yml'), 'w') as f_world:
-            f_world.write(ROLE_TASKS_WORLD)
-
-    def tearDown(self):
-        shutil.rmtree(self.play_root)
-
-    def _get_play_file(self, playbook_text):
-        with open(os.path.join(self.play_root, 'playbook.yml'), 'w') as f_play:
-            f_play.write(playbook_text)
-        return f_play
-
-    def test_import_role(self):
-        fh = self._get_play_file(PLAY_IMPORT_ROLE)
-        runner = Runner(self.rules, fh.name, [], [], [])
-        results = runner.run()
-        assert 'only when shell functionality is required' in str(results)
-
-    def test_import_role_inline_args(self):
-        fh = self._get_play_file(PLAY_IMPORT_ROLE_INLINE)
-        runner = Runner(self.rules, fh.name, [], [], [])
-        results = runner.run()
-        assert 'only when shell functionality is required' in str(results)
-
-    def test_include_role(self):
-        fh = self._get_play_file(PLAY_INCLUDE_ROLE)
-        runner = Runner(self.rules, fh.name, [], [], [])
-        results = runner.run()
-        assert 'only when shell functionality is required' in str(results)
-        assert 'All tasks should be named' in str(results)
-
-    def test_include_role_inline_args(self):
-        fh = self._get_play_file(PLAY_INCLUDE_ROLE_INLINE)
-        runner = Runner(self.rules, fh.name, [], [], [])
-        results = runner.run()
-        assert 'only when shell functionality is required' in str(results)
-        assert 'All tasks should be named' in str(results)
-
-    def test_invalid_import_role(self):
-        fh = self._get_play_file(PLAY_IMPORT_ROLE_INCOMPLETE)
-        runner = Runner(self.rules, fh.name, [], [], [])
-        with pytest.raises(RuntimeError, match="Failed to find required 'name' key in import_role"):
-            runner.run()
+def _setup_play(tmpdir, playbook_text):
+    # make dir and write role tasks to import or include
+    role_path = os.path.join(tmpdir, 'test-role', 'tasks')
+    os.makedirs(role_path)
+    with open(os.path.join(role_path, 'main.yml'), 'w') as f_main:
+        f_main.write(ROLE_TASKS_MAIN)
+    with open(os.path.join(role_path, 'world.yml'), 'w') as f_world:
+        f_world.write(ROLE_TASKS_WORLD)
+
+    play_path = os.path.join(tmpdir, 'playbook.yml')
+    with open(play_path, 'w') as f_play:
+        f_play.write(playbook_text)
+    return play_path
```

Comment:
FYI tmpdir is a pathlib.Path-like object. So instead of converting things to old-fashioned strings, you should use its interface:

```suggestion
    role_tasks_dir = tmpdir / 'test-role' / 'tasks'
    role_tasks_dir.mkdir(parents=True)

    (role_tasks_dir / 'main.yml').write_text(ROLE_TASKS_MAIN)
    (role_tasks_dir / 'world.yml').write_text(ROLE_TASKS_WORLD)

    play_path = tmpdir / 'playbook.yml'
    play_path.write_text(playbook_text)
    return play_path
```

Ref: https://docs.pytest.org/en/latest/tmpdir.html

Reply:
Yes and No. Changelog says yes, since pytest 3.9, then around 4.x it seems to revert back to something else (referred to as `py.path.local`). On my computer with PyTest 5.2.2 / Py3.7 this is not a `pathlib.Path`, it is a `py._path.local.LocalPath` (displayed as `local('<path>')`) and that does not behave like a `Path` (_e.g._ its `mkdir()` method does not support the `parents` keyword argument). I guess there is some portability stuff still around in PyTest's code. Maybe the situation will become clearer in the future.

In the meantime, If I switch to `tmp_path`, which is a `pathlib.Path`, then later on the playbook fails to load. It happens somewhere in Ansible's API which, for portability's sake we have to avoid. So I went for the swtich to `tmp_path` + `return str(play_path)` on the last statement. Not as elegant as what you suggested, but hopefully close enough.

Traceback relative to the ansible issue:
```
test/TestImportIncludeRole.py:124: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
lib/ansiblelint/__init__.py:266: in run
    for child in ansiblelint.utils.find_children(arg, self.playbook_dir):
lib/ansiblelint/utils.py:156: in find_children
    playbook_ds = parse_yaml_from_file(playbook[0])
lib/ansiblelint/utils.py:60: in parse_yaml_from_file
    return dl.load_from_file(filepath)
../../../../.virtualenvs/ansible-lint/lib/python3.7/site-packages/ansible/parsing/dataloader.py:85: in load_from_file
    file_name = self.path_dwim(file_name)
../../../../.virtualenvs/ansible-lint/lib/python3.7/site-packages/ansible/parsing/dataloader.py:186: in path_dwim
    given = unquote(given)
../../../../.virtualenvs/ansible-lint/lib/python3.7/site-packages/ansible/parsing/quoting.py:29: in unquote
    if is_quoted(data):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

data = PosixPath('/tmp/pytest-of-user/pytest-47/test_invalid_import_role_IMPOR0/playbook.yml')

    def is_quoted(data):
>       return len(data) > 1 and data[0] == data[-1] and data[0] in ('"', "'") and data[-2] != '\\'
E       TypeError: object of type 'PosixPath' has no len()

data       = PosixPath('/tmp/pytest-of-user/pytest-47/test_invalid_import_role_IMPOR0/playbook.yml')

../../../../.virtualenvs/ansible-lint/lib/python3.7/site-packages/ansible/parsing/quoting.py:24: TypeError
```

Reply:
Well, in this case, I'd do `str(path)` when passing the value to the library calls but keep the pathlib object where you can control it. But I guess we could skip this for now.

Reply:
https://github.com/ansible/ansible-lint/pull/740#issuecomment-617896072